timApp/static/scripts/tim/plugin/quantumcircuit/quantum-board.ts

```diff
@@ -0,0 +1,556 @@
+            const cell = this.get(gatePos.target, gatePos.time);
+            if (cell instanceof Gate) {
+                this.set(
+                    pos.target,
+                    pos.time,
+                    new Control(gatePos.target, true)
+                );
+            }
+        } else if (!gatePos) {
+            const target = this.findControllable(pos);
+            if (target !== undefined) {
+                this.set(pos.target, pos.time, new Control(target, true));
+            }
+        }
+    }
+
+    /**
+     * Find empty cell in the time slot as given position that
+     * isn't the position given.
+     * @param pos position to look swap pair for
+     */
+    private findSwapPair(pos: GatePos) {
+        // Prefer positions after current position
+        for (let i = pos.target + 1; i < this.board.length; i++) {
+            if (!this.board[i][pos.time]) {
+                return i;
+            }
+        }
+        // then before in reverse order
+        for (let i = pos.target - 1; i >= 0; i--) {
+            if (!this.board[i][pos.time]) {
+                return i;
+            }
+        }
+        return undefined;
+    }
+
+    /**
+     * Adds swap gate between two positions.
+     * Automatically selects pair for it from available empty cells
+     * if pos2 is not given.
+     * Removes connections to gate at position.
+     * @param pos position to add swap to
+     * @param pos2 position of the pair of the swap gate
+     * @param editable whether the swap gate is editable
+     */
+    addSwap(pos: GatePos, pos2: GatePos | null = null, editable: boolean) {
+        if (this.get(pos.target, pos.time)?.editable === false) {
+            return;
+        }
+        if (
+            pos2 !== null &&
+            this.get(pos2.target, pos2.time)?.editable === false
+        ) {
+            return;
+        }
+        if (pos.target === pos2?.target) {
+            return;
+        }
+        if (pos2 && pos.time === pos2.time) {
+            this.remove(pos.target, pos.time);
+            this.remove(pos2.target, pos2.time);
+            this.board[pos.target][pos.time] = new Swap(pos2.target, editable);
+            this.board[pos2.target][pos2.time] = new Swap(pos.target, editable);
+        } else {
+            const pair = this.findSwapPair(pos);
+            if (pair !== undefined) {
+                this.remove(pos.target, pos.time);
+                this.remove(pair, pos.time);
+                this.board[pos.target][pos.time] = new Swap(pair, editable);
+                this.board[pair][pos.time] = new Swap(pos.target, editable);
+            }
+        }
+    }
+
+    /**
+     * Sets cell in board.
+     * @param target qubit associated with cell
+     * @param time time of cell
+     * @param value new value for cell
+     */
+    set(target: number, time: number, value: Cell) {
+        if (this.isInvalidPosition(target, time)) {
+            console.log("invalid indices", target, time);
```

Comment:
Tähän voisi miettiä vähän virheenkäsittelyä. Tällä hetkellä tämä lokittaa muttei tule `return`ia. Vastaavia kohtia on muuallakin koodissa.

Reply:
laitoin returnit noihin. Nuo virheelliset paikat taitaa tulla tällä hetkellä vain initialCircuit:n kautta, mutta niiden käsittelyyn tarviis vielä lisää tarkistuksia. Pitäis kattoo mitä tuo Qulacs tekee noiden virheiden tapauksessa, jos se vaikka heittäis poikkeuksia ni pääsis helpolla palvelinpuolella